### PR TITLE
Add ActiveDevices to statistics and other updates

### PR DESCRIPTION
**Pull request by avanikp**
# Description and Motivation 

- Add ActiveDevices to simulation stats (read/write data to stats records).
- Update AvgMessageCount to return value rounded to 2 decimal places.
- Add stats to entire list of simulations in GetAllAsync.
- Update CreateOrUpdateAsync to call Upsert in case the record already exists.
- Add null checks for simulation actual start time and stopped time.
- Add/Update Unit tests and doc.

# Change type 

- [X] Bug fix
- [ ] New feature
- [X] Enhancement
- [ ] Breaking change (breaks backward compatibility)

**Checklist:**

- [ ] All tests passed
- [ ] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

---
This change is [Reviewable](https://reviewable.io/reviews/azure/device-simulation-dotnet/304)

### CODE REVIEW

File: WebService/v1/Models/SimulationApiModel/SimulationApiModel.cs [hunk 1]
```diff
@@ -268,11 +268,22 @@ private static SimulationStatistics GetSimulationStatistics(Simulation simulatio
...
+                {
+                    durationInSeconds = now.Subtract(actualStartTime).TotalSeconds;
+                }
+                else if (simulation.StoppedTime.HasValue)
+                {
+                    durationInSeconds = simulation.StoppedTime.Value.Subtract(actualStartTime).TotalSeconds;
+                }
+
+                statistics.AverageMessagesPerSecond = durationInSeconds > 0 ? Math.Round((double) statistics.TotalMessagesSent / durationInSeconds, 2) : 0;
             }
```
Comment (hathind-ms) [inline WebService/v1/Models/SimulationApiModel/SimulationApiModel.cs]: Use Math. ceil instead:

Math.Ceiling( statistics.TotalMessagesSent / durationInSeconds* 100) / 100

File: Services/Statistics/SimulationStatistics.cs [hunk 2]
```diff
@@ -91,8 +92,17 @@ public async Task CreateOrUpdateAsync(string simulationId, SimulationStatisticsM
 
             try
             {
-                this.log.Debug("Creating statistics record", () => new { statisticsStorageRecord });
-                await this.simulationStatisticsStorage.CreateAsync(statisticsStorageRecord);
+                if (await this.simulationStatisticsStorage.ExistsAsync(statisticsRecordId))
```
Comment (peterfelts) [inline Services/Statistics/SimulationStatistics.cs]: I would try to get the existing record, if it exists, and then pass the ETag. If it doesn't exist, you could pass a wildcard for the ETag. The storage library will handle this and will do the right thing with the DocumentDB API.

`var ETag = "*";
var existingRecord = await this.simulationStatisticsStorage.ExistsAsync(statisticsRecordId);
if(existingRecord != null) ETag = existingRecord.ETag;
await this.simulationStatisticsStorage.UpsertAsync(statisticsStorageRecord, ETag);`

File: WebService/v1/Models/SimulationApiModel/SimulationApiModel.cs (hunk 1 above)
Comment (avanikp) [reply WebService/v1/Models/SimulationApiModel/SimulationApiModel.cs]: If we want value upto 2 decimal places, using Math.Round should give the expected value. I checked the value returned in both cases.

File: Services/Statistics/SimulationStatistics.cs (hunk 2 above)
Comment (peterfelts) [reply Services/Statistics/SimulationStatistics.cs]: According to the [DocumentDB documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.requestoptions?view=azure-dotnet), you can send a meaningless ETag for creation; it'll be ignored. 

"ETag, while valid on Replace* and Delete* operations, would have no impact on a Read*, CreateQuery* or Create* operations."

File: Services/Simulations.cs [hunk 3]
```diff
@@ -21,6 +21,9 @@ public interface ISimulations
         // Get list of simulations.
         Task<IList<Models.Simulation>> GetListAsync();
 
+        // Get list of simulations with statistics.
+        Task<IList<Models.Simulation>> GetListWithStatisticsAsync();
```
Comment (peterfelts) [inline Services/Simulations.cs]: Please add a comment explaining that GetList has better performance than calling GetListWithStatisticsAsync.

File: WebService/v1/Models/SimulationApiModel/SimulationApiModel.cs [hunk 4]
```diff
@@ -268,11 +268,22 @@ private static SimulationStatistics GetSimulationStatistics(Simulation simulatio
         {
             var statistics = SimulationStatistics.FromServiceModel(simulation.Statistics);
 
-            if (statistics != null)
+            if (statistics != null && simulation.ActualStartTime.HasValue)
```
Comment (peterfelts) [inline WebService/v1/Models/SimulationApiModel/SimulationApiModel.cs]: Please add a comment explaining why you're checking for simulations that have a value for ActualStartTime. What happens if I pass in a simulation that isn't running?

Comment (avanikp) [reply WebService/v1/Models/SimulationApiModel/SimulationApiModel.cs]: sure added

File: Services/Simulations.cs (hunk 3 above)
Comment (avanikp) [reply Services/Simulations.cs]: added

File: Services/Statistics/SimulationStatistics.cs (hunk 2 above)
Comment (avanikp) [reply Services/Statistics/SimulationStatistics.cs]: Good point about using Upsert for both cases.  In both cases we have similar checks for Exists and pass expected eTag and also it seems cleaner to call Create if the record does not exist. I prefer to keep this implementation.

**Pull request merged**